From a83813fb401c29a9a2eff6b7140c1ee96d278afd Mon Sep 17 00:00:00 2001
From: Nicolas Aguirre <nicolas.aguirre@openwide.fr>
Date: Wed, 15 Apr 2015 12:56:51 +0200
Subject: [PATCH] decrease pcie_set_read_rq size from 4096 to 256

---
 drivers/ata/sata_sil24.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/ata/sata_sil24.c b/drivers/ata/sata_sil24.c
index 59f0d63..e728b8e 100644
--- a/drivers/ata/sata_sil24.c
+++ b/drivers/ata/sata_sil24.c
@@ -1336,7 +1336,7 @@ static int sil24_init_one(struct pci_dev *pdev, const struct pci_device_id *ent)
 	/* Set max read request size to 4096.  This slightly increases
 	 * write throughput for pci-e variants.
 	 */
-	pcie_set_readrq(pdev, 4096);
+	pcie_set_readrq(pdev, 256);
 
 	sil24_init_controller(host);
 
-- 
1.9.1

