diff --git a/arch/arm/boot/dts/imx6qdl-sabrelite.dtsi b/arch/arm/boot/dts/imx6qdl-sabrelite.dtsi
index 46f6fc4..d4a7515 100644
--- a/arch/arm/boot/dts/imx6qdl-sabrelite.dtsi
+++ b/arch/arm/boot/dts/imx6qdl-sabrelite.dtsi
@@ -461,6 +461,18 @@
                 wakeup-gpios = <&gpio1 9 0>;
         };

+        ct363_ts@01 {
+                compatible = "ct363_ts,ct363_ts";
+                reg = <0x01>;
+                ct-model = <363>;
+                touch-gpio = <&gpio1 9 0>;
+                reset-gpio = <&gpio2 0 GPIO_ACTIVE_HIGH>;
+                max-x = <1024>;
+                max-y = <768>;
+                orientation = <1 0 0 1>;
+        };
+
+
         ft5x06_ts@38 {
 		compatible = "ft5x06-ts,ft5x06-ts";
 		reg = <0x38>;